lib/view_component/compiler.rb

```diff
@@ -103,266 +50,273 @@ def renders_template_for?(variant, format)
         def render_template_for(variant = nil, format = nil)
-          #{body}
+          #{method_body}
         end
         RUBY
       end
     end
 
-    def has_inline_template?
-      component_class.respond_to?(:inline_template) && component_class.inline_template.present?
-    end
+    def gather_template_errors(raise_errors)
+      errors = []
 
-    def template_errors
-      @__vc_template_errors ||=
-        begin
-          errors = []
+      errors << "Couldn't find a template file or inline render method for #{@component}." if @templates.empty?
 
-          if (templates + inline_calls).empty? && !has_inline_template?
-            errors << "Couldn't find a template file or inline render method for #{component_class}."
-          end
+      # We currently allow components to have both an inline call method and a template for a variant, with the
+      # inline call method overriding the template. We should aim to change this in v4 to instead
+      # raise an error.
+      @templates.select { !_1.inline_call? }
+        .map { |template| [template.variant, template.format] }
+        .tally
+        .select { |_, count| count > 1 }
+        .each do |tally|
+        variant, this_format = tally.first
 
-          templates
-            .map { |template| [template[:variant], template[:format]] }
-            .tally
-            .select { |_, count| count > 1 }
-            .each do |tally|
-            variant, this_format = tally[0]
+        variant_string = " for variant `#{variant}`" if variant.present?
 
-            variant_string = " for variant `#{variant}`" if variant.present?
+        errors << "More than one #{this_format.upcase} template found#{variant_string} for #{@component}. "
+      end
 
-            errors << "More than one #{this_format.upcase} template found#{variant_string} for #{component_class}. "
-          end
+      if @templates.any? { _1.variant.nil? && !_1.inline_call? } &&
+          @templates.any? { _1.variant.nil? && _1.inline_call? && _1.defined_on_self? }
 
-          if templates.find { |template| template[:variant].nil? } && inline_calls_defined_on_self.include?(:call)
-            errors <<
-              "Template file and inline render method found for #{component_class}. " \
-              "There can only be a template file or inline render method per component."
-          end
+        errors <<
+          "Template file and inline render method found for #{@component}. " \
+          "There can only be a template file or inline render method per component."
+      end
 
-          duplicate_template_file_and_inline_variant_calls =
-            templates.pluck(:variant) & variants_from_inline_calls(inline_calls_defined_on_self)
+      duplicate_template_file_and_inline_call_variants =
+        @templates.select { !_1.inline_call? }.map(&:variant) &
+        @templates.select { _1.inline_call? && _1.defined_on_self? }.map(&:variant)
```

Comment:
```suggestion
        # If a template has inline calls, they can conflict with template files the component may use
        # to render. This attempts to catch and raise that issue before run time. For example,
        # `def render_mobile` would conflict with a sidecar template of `component.html+mobile.erb`
        @templates.select(&:variant) & @templates.select { |template| template.inline_call? && template.defined_on_self? }.map(&:variant)
```

Thoughts on adding a helpful comment here?